controllers/circuitrelay_controller.go

```diff
@@ -0,0 +1,388 @@
+
+	var requeue bool
+	for obj, mut := range trackedObjects {
+		kind := obj.GetObjectKind().GroupVersionKind()
+		name := obj.GetName()
+		result, err := controllerutil.CreateOrPatch(ctx, r.Client, obj, mut)
+		if err != nil {
+			log.Error(err, "error creating object", "objname", name, "objKind", kind.Kind, "result", result)
+			requeue = true
+		} else {
+			log.Info("object changed", "objName", name, "objKind", kind.Kind, "result", result)
+		}
+	}
+	return ctrl.Result{Requeue: requeue}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CircuitRelayReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
+		For(&clusterv1alpha1.CircuitRelay{}).
+		Owns(&corev1.Service{}, builder.OnlyMetadata).
+		Owns(&appsv1.Deployment{}, builder.OnlyMetadata).
+		Complete(r)
+}
+
+func (r *CircuitRelayReconciler) serviceRelay(m *clusterv1alpha1.CircuitRelay, svc *corev1.Service) controllerutil.MutateFn {
+	svcName := "libp2p-relay-daemon-" + m.Name
+	expected := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: m.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeLoadBalancer,
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "swarm",
+					Protocol:   corev1.ProtocolTCP,
+					Port:       4001,
+					TargetPort: intstr.FromString("swarm"),
+				},
+				// Commented out because some load balancers don't support TCP+UDP:(
+				//
+				// {
+				// 	Name:       "swarm-udp",
+				// 	Protocol:   corev1.ProtocolUDP,
+				// 	Port:       4001,
+				// 	TargetPort: intstr.FromString("swarm-udp"),
+				// },
+				//
+				// Commented out.
+				// TODO: support websockets and TLS.
+				// {
+				// 	Name:       "swarm-ws",
+				// 	Protocol:   corev1.ProtocolTCP,
+				// 	Port:       8081,
+				// 	TargetPort: intstr.FromString("swarm-ws"),
+				// },
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name": svcName,
+			},
+		},
+	}
+	expected.DeepCopyInto(svc)
+	ctrl.SetControllerReference(m, svc, r.Scheme)
+	return func() error {
+		svc.Spec = expected.Spec
+		return nil
+	}
+}
+
+func (r *CircuitRelayReconciler) secretIdentity(m *clusterv1alpha1.CircuitRelay, sec *corev1.Secret, identity []byte) controllerutil.MutateFn {
+	secName := "libp2p-relay-daemon-identity-" + m.Name
+	expected := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secName,
+			Namespace: m.Namespace,
+		},
+		Data: map[string][]byte{
+			"identity": identity,
+		},
+	}
+	expected.DeepCopyInto(sec)
+	ctrl.SetControllerReference(m, sec, r.Scheme)
+	return func() error {
+		return nil
+	}
+}
+
+func (r *CircuitRelayReconciler) configRelay(m *clusterv1alpha1.CircuitRelay, cm *corev1.ConfigMap) controllerutil.MutateFn {
+	cmName := "libp2p-relay-daemon-config-" + m.Name
+	m.Status.AddrInfo.Parse()
```

Comment:
Error should be checked here

Reply:
Right you are. Will fix that.

Reply:
Actually, as you pointed out in another comment, probably don't need to re-parse this.